### PR TITLE
feat(kuma-cp) refactor SDS to go-control-plane building blocks

### PR DESCRIPTION
**Pull request by jakubdyszkiewicz**
### Summary

This PR refactors SDS to use the same patterns as our XDS implementation.

Until now, SDS was custom implementation operating on raw Envoy protos.
The problem with this implementation is that
1) It's hard to build logic on top of it
2) We generate a certificate on every stream request. Example:
DP has 1000 outbound listeners. Since it is mTLS, we have to configure every outbound listener with DP cert and CA cert. Every DP will open 2000 streams (one for CA, one for DP cert) and send a request on every stream. This will generate 1000 certs for one DP instead of one.

Given that there can be thousands of such dataplanes and CA can be external system, we can overload the system with generating millions of certs.

In this implementation we use SnapshotCache from go-control-plane which will store the same cert for subsequent requests, so we only generate 1 certificate per Dataplane.

This will also helps us to solve #719

### CODE REVIEW

File: pkg/sds/server/auth_callbacks.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	}
+}
+
+// AuthCallback checks if the DiscoveryRequest is authorized, ie. if it has a valid Dataplane Token/Service Account Token.
+type authCallbacks struct {
+	authenticator sds_auth.Authenticator
+
+	mutex    sync.RWMutex
+	contexts map[int64]context.Context
+}
```
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/sds/server/auth_callbacks.go:29]: Done

File: pkg/plugins/resources/memory/store.go [hunk 2]
```diff
@@ -146,6 +146,8 @@ func (c *memoryStore) Update(_ context.Context, r model.Resource, fs ...store.Up
 
 	// persist
 	c.records[idx] = record
+
+	r.SetMeta(meta)
```
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/plugins/resources/memory/store.go:150]: Actually it was a bug in memory store. We should update meta of the object that we are passing. We do this in postgres and K8S. I added a test in `store template` to cover this for all stores. I discovered also small bug in postgres (we update meta but leave old version)

File: pkg/sds/server/auth_callbacks.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	"github.com/pkg/errors"
+
+	core_xds "github.com/Kong/kuma/pkg/core/xds"
+	sds_auth "github.com/Kong/kuma/pkg/sds/auth"
+)
+
+func newAuthCallbacks(authenticator sds_auth.Authenticator) *authCallbacks {
+	return &authCallbacks{
+		authenticator: authenticator,
+		mutex:         sync.RWMutex{},
```
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/sds/server/auth_callbacks.go]: Good point, removed and mutex is now embedded

File: pkg/sds/server/auth_callbacks.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+	envoy_api "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	envoy_server "github.com/envoyproxy/go-control-plane/pkg/server/v2"
+	"github.com/pkg/errors"
+
+	core_xds "github.com/Kong/kuma/pkg/core/xds"
+	sds_auth "github.com/Kong/kuma/pkg/sds/auth"
+)
+
+func newAuthCallbacks(authenticator sds_auth.Authenticator) *authCallbacks {
```
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/sds/server/auth_callbacks.go]: We can and we should

File: pkg/sds/server/auth_callbacks.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+		contexts:      map[int64]context.Context{},
+	}
+}
+
+// AuthCallback checks if the DiscoveryRequest is authorized, ie. if it has a valid Dataplane Token/Service Account Token.
+type authCallbacks struct {
+	authenticator sds_auth.Authenticator
+
+	mutex    sync.RWMutex
+	contexts map[int64]context.Context
```
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/sds/server/auth_callbacks.go]: Added alias

File: pkg/sds/server/reconciler.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+func (d *DataplaneReconciler) Reconcile(dataplaneId core_model.ResourceKey) error {
+	proxyID := core_xds.FromResourceKey(dataplaneId).String()
+
+	dataplane := &mesh_core.DataplaneResource{}
+	if err := d.resManager.Get(context.Background(), dataplane, core_store.GetBy(dataplaneId)); err != nil {
+		if core_store.IsResourceNotFound(err) {
+			sdsServerLog.V(1).Info("Dataplane not found. Clearing the Snapshot.", "dataplaneId", dataplaneId)
+			d.cache.ClearSnapshot(proxyID)
```
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/sds/server/reconciler.go:44]: I intentionally left this so the error is thrown, but now that I think about this, it makes more sense to return from here

File: pkg/sds/server/reconciler.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+		}
+		return err
+	}
+
+	meshRes := mesh_core.MeshResource{}
```
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/sds/server/reconciler.go]: good point, changed

File: pkg/sds/server/reconciler.go [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+func (d *DataplaneReconciler) shouldGenerateSnapshot(proxyID string, meshRes mesh_core.MeshResource) (bool, error) {
+	currentSnapshot, err := d.cache.GetSnapshot(proxyID)
+	if err != nil { // snapshot does not exist
+		return true, nil
+	}
+
+	parts := strings.Split(currentSnapshot.GetVersion(envoy_resource.SecretType), "-")
+	if len(parts) != 2 {
+		return false, errors.New(`invalid snapshot version format. Format should be "UnixNano-NameOfTheCA"`)
```
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/sds/server/reconciler.go:87]: Not really, unless someone connects to the control plane and explicitly set the version. Or we change it in the future and it won't be compatible with current format.

But let's say it should not happen. What is the alternative? Panic or ignore the error?

[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/sds/server/reconciler.go:87]: If we want to protect ourselves from our future changes in the format, then it feels more like `panic`. But that's really minor comment, so up to you :)

**Pull request merged**